src/PsrLogAdapter.php

```diff
@@ -0,0 +1,118 @@
+use MongoDB\Exception\UnexpectedValueException;
+use Psr\Log\LoggerInterface as PsrLogger;
+use Psr\Log\LogLevel as PsrLogLevel;
+use SplObjectStorage;
+
+use function MongoDB\Driver\Logging\addLogger;
+use function MongoDB\Driver\Logging\removeLogger;
+use function sprintf;
+
+final class PsrLogAdapter implements DriverLogger
+{
+    private static ?self $instance = null;
+
+    /** @psalm-var SplObjectStorage<PsrLogger, null> */
+    private SplObjectStorage $psrLoggers;
+
+    private const PSR_LEVELS = [
+        DriverLogger::LEVEL_ERROR => PsrLogLevel::ERROR,
+        /* libmongoc considers "critical" less severe than "error" so map it to
+         * "error" in the PSR logger. */
+        DriverLogger::LEVEL_CRITICAL => PsrLogLevel::ERROR,
+        DriverLogger::LEVEL_WARNING => PsrLogLevel::WARNING,
+        DriverLogger::LEVEL_MESSAGE => PsrLogLevel::NOTICE,
+        DriverLogger::LEVEL_INFO => PsrLogLevel::INFO,
+        DriverLogger::LEVEL_DEBUG => PsrLogLevel::DEBUG,
+    ];
+
+    public static function addLogger(PsrLogger $psrLogger): void
+    {
+        $instance = self::getInstance();
+
+        $instance->psrLoggers->attach($psrLogger);
+
+        addLogger($instance);
+    }
+
+    /**
+     * Pipes a log message from PHPC to all registered PSR loggers.
+     *
+     * @internal
+     * @see DriverLogger::log()
+     */
+    public function log(int $level, string $domain, string $message): void
+    {
+        if (! isset(self::PSR_LEVELS[$level])) {
+            throw new UnexpectedValueException(sprintf(
+                'Expected level to be >= %d and <= %d, %d given for domain "%s" and message: %s',
+                DriverLogger::LEVEL_ERROR,
+                DriverLogger::LEVEL_DEBUG,
+                $level,
+                $domain,
+                $message,
+            ));
+        }
+
+        $instance = self::getInstance();
+        $psrLevel = self::PSR_LEVELS[$level];
+        $context = ['domain' => $domain];
+
+        foreach ($instance->psrLoggers as $psrLogger) {
+            $psrLogger->log($psrLevel, $message, $context);
+        }
+    }
+
+    public static function removeLogger(PsrLogger $psrLogger): void
+    {
+        $instance = self::getInstance();
+        $instance->psrLoggers->detach($psrLogger);
+
+        if ($instance->psrLoggers->count() === 0) {
+            removeLogger($instance);
+        }
+    }
+
+    /**
+     * Writes a log message to all registered PSR loggers.
+     *
+     * @internal
+     */
+    public static function writeLog(int $level, string $domain, string $message): void
+    {
+        self::getInstance()->log($level, $domain, $message);
+    }
+
+    private function __construct()
+    {
+        $this->psrLoggers = new SplObjectStorage();
+    }
+
+    private static function getInstance(): self
+    {
+        if (self::$instance === null) {
+            self::$instance = new self();
+        }
+
+        return self::$instance;
```

Comment:
There is an operator for that.
```suggestion
        return self::$instance ??= new self();
```

Reply:
Thanks. I forgot that was introduced in PHP 7.4. :v: